### PR TITLE
osd: Implement lazy omap usage statistics

### PR DESCRIPTION
**Pull request by badone**
Opportunistic gathering of omap statistics during deep scrub.

Signed-off-by: Brad Hubbard <bhubbard@redhat.com>

- [ ] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by badone** [comment]
Currently collection and display of omap statistics gathered during deep scrub are enabled by this commit but these stats are potentially inaccurate before they are completely gathered. Therefore, I wanted to get some feedback on how we relay that to the end user. We could use the last_deep_scrub_stamp as a timestamp and implement a function to invalidate the results based on that but what does an invalidated result look like in the output and how does it differ from a zero result (if at all). We could also just document the fact that to get an accurate output you need to run a deep scrub first. Thoughts?

**Comment by neha-ojha** [comment]
@badone I think it is important to correctly inform users that the stats they are looking at could be invalid, beyond just documenting it. So, either 1) we do not report them at all when that's the case or 2) mark the stats `inaccurate` when we display them and recommend a deep-scrub for accurate stats.

**Comment by badone** [comment]
@neha-ojha Right, but I was wondering specifically how long stats stay valid (say an arbitrary value of 6 hours) and what an invalid stat looks like in the output (somehow "tagged" as invalid). I was trying to avoid having to code this up and then being requested to change it but I think that's probably going to end up being the "quickest" way...

**Comment by gregsfortytwo** [comment]
Well the stats are pretty much always going to be inaccurate, unless you manage to complete a deep scrub on a PG without getting *any* map updates to it during that time.

So I guess we could just make it part of every pg dump — stick an asterisk next to the column and print out a short explanation at the bottom, with a link to a longer doc page? I don't really see how else to handle it besides writing a very clear explanation and trying to make it apparent on every output that something's at least a little bit kinky about them.

**Comment by badone** [comment]
@gregsfortytwo Sure, that's reasonable. I don't think we should create any hard url links though because that could be difficult to maintain in different versions? Maybe just "search the documentation for "lazy omap" or something like that? Thanks.

**Comment by gregsfortytwo** [comment]
No we can definitely have a permanent URL link, and AFAIK that's considered good practice for serious error messages at this point. Our docs URL scheme hasn't really ever changed that I recall.

**Comment by badone** [comment]
OK. What say you now?

**Comment by gregsfortytwo** [comment]
Looks fine in code, but do you have an example pg dump output to show off? :)

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -3273,15 +3302,20 @@ int process_pg_map_command(
...
 	  pg_map.dump_pool_stats(ds, header);
+          omap_stats_note_required = true;
 	}
 	if (what.count("osds")) {
 	  pg_map.dump_osd_stats(ds);
 	}
       }
       odata->append(ds);
+      if (omap_stats_note_required) {
+        odata->append(omap_stats_note);
```
Comment (gregsfortytwo) [inline src/mon/PGMap.cc]: Not a big deal, but maybe this should append to ds prior to the odata append?

**Comment by gregsfortytwo** [comment]
Oh, and some kind of test? Not sure what the best option is, maybe just shove some known-size omap data in and force a deep scrub and see it appear in part of the standalone suite scripts?

**Comment by badone** [comment]
```
$ ceph pg dump
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-02-27 11:49:54.497 7fd171a53700 -1 WARNING: all dangerous and experimental features are enabled.
2019-02-27 11:49:54.667 7fd171a53700 -1 WARNING: all dangerous and experimental features are enabled.
dumped all
version 20172
stamp 2019-02-27 11:49:53.297408
last_osdmap_epoch 0
last_pg_scan 0
PG_STAT OBJECTS MISSING_ON_PRIMARY DEGRADED MISPLACED UNFOUND BYTES OMAP_BYTES* OMAP_KEYS* LOG DISK_LOG STATE        STATE_STAMP                VERSION REPORTED UP      UP_PRIMARY ACTING  ACTING_PRIMARY LAST_SCRUB SCRUB_STAMP                LAST_DEEP_SCRUB DEEP_SCRUB_STAMP           SNAPTRIMQ_LEN 
1.9          16                  0        0         0       0     0   201326592         48  16       16 active+clean 2019-02-27 07:08:43.559261   13'16    16:40 [1,2,0]          1 [1,2,0]              1      13'16 2019-02-26 14:22:42.286676           13'16 2019-02-26 14:22:42.286676             0 
1.8          21                  0        0         0       0     0   264241152         63  21       21 active+clean 2019-02-27 07:08:43.554860   13'21    16:45 [1,2,0]          1 [1,2,0]              1      13'21 2019-02-26 14:22:41.355824           13'21 2019-02-26 14:22:41.355824             0 
1.7          39                  0        0         0       0     0   490733568        117  39       39 active+clean 2019-02-27 07:08:43.545804   13'39    16:63 [1,2,0]          1 [1,2,0]              1      13'39 2019-02-26 14:22:40.473369           13'39 2019-02-26 14:22:40.473369             0 
1.6          35                  0        0         0       0     0   440401920        105  35       35 active+clean 2019-02-27 07:08:43.556879   13'35    16:59 [1,0,2]          1 [1,0,2]              1      13'35 2019-02-26 14:22:36.503362           13'35 2019-02-26 14:22:36.503362             0 
1.1          28                  0        0         0       0     0   352321536         84  28       28 active+clean 2019-02-27 07:08:43.535766   13'28    16:52 [2,0,1]          2 [2,0,1]              2      13'28 2019-02-26 14:22:23.320667           13'28 2019-02-26 14:22:23.320667             0 
1.0          23                  0        0         0       0     0   289406976         69  23       23 active+clean 2019-02-27 07:08:43.559662   13'23    16:47 [1,0,2]          1 [1,0,2]              1      13'23 2019-02-26 14:22:30.478020           13'23 2019-02-26 14:22:30.478020             0 
1.2          38                  0        0         0       0     0   478150656        114  38       38 active+clean 2019-02-27 07:08:43.539575   13'38    16:62 [0,1,2]          0 [0,1,2]              0      13'38 2019-02-26 14:22:24.268516           13'38 2019-02-26 14:22:24.268516             0 
1.3          27                  0        0         0       0     0   339738624         81  27       27 active+clean 2019-02-27 07:08:43.552369   13'27    16:51 [1,2,0]          1 [1,2,0]              1      13'27 2019-02-26 14:22:31.512259           13'27 2019-02-26 14:22:31.512259             0 
1.4          32                  0        0         0       0     0   402653184         96  32       32 active+clean 2019-02-27 07:08:43.554496   13'32    16:56 [1,0,2]          1 [1,0,2]              1      13'32 2019-02-26 14:22:35.549114           13'32 2019-02-26 14:22:35.549114             0 
1.5          37                  0        0         0       0    24   452984832        108  37       37 active+clean 2019-02-27 07:08:43.536722   13'37    16:61 [2,0,1]          2 [2,0,1]              2      13'37 2019-02-26 14:22:26.340739           13'37 2019-02-26 14:22:26.340739             0 
                                        
1 296 0 0 0 0 24 3711959040 885 296 296 
                                          
sum 296 0 0 0 0 24 3711959040 885 296 296 
OSD_STAT USED    AVAIL   USED_RAW TOTAL  HB_PEERS PG_SUM PRIMARY_PG_SUM 
2         65 MiB 8.9 GiB  1.2 GiB 10 GiB    [0,1]     10              2 
0         65 MiB 8.9 GiB  1.2 GiB 10 GiB    [1,2]     10              1 
1         65 MiB 8.9 GiB  1.2 GiB 10 GiB    [0,2]     10              7 
sum      194 MiB  27 GiB  3.5 GiB 30 GiB                                

* NOTE: Omap statistics are gathered during deep scrub and may be inaccurate soon afterwards depending on utilisation. See http://docs.ceph.com/docs/master/dev/placement-group/#omap-statistics for further details.
```

**Comment by badone** [comment]
```
$ ceph pg dump summary
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-02-27 11:51:24.686 7f304d87a700 -1 WARNING: all dangerous and experimental features are enabled.
2019-02-27 11:51:24.854 7f304d87a700 -1 WARNING: all dangerous and experimental features are enabled.
dumped summary
version 20217
stamp 2019-02-27 11:51:23.375252
last_osdmap_epoch 0
last_pg_scan 0
PG_STAT OBJECTS MISSING_ON_PRIMARY DEGRADED MISPLACED UNFOUND BYTES OMAP_BYTES* OMAP_KEYS* LOG DISK_LOG 
sum         296                  0        0         0       0    24  3711959040        885 296      296 
OSD_STAT USED    AVAIL  USED_RAW TOTAL  
sum      194 MiB 27 GiB  3.5 GiB 30 GiB 

* NOTE: Omap statistics are gathered during deep scrub and may be inaccurate soon afterwards depending on utilisation. See http://docs.ceph.com/docs/master/dev/placement-group/#omap-statistics for further details.
```

**Comment by badone** [comment]
```
$ ceph pg dump pgs
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-02-27 11:52:02.407 7f2c6ed7a700 -1 WARNING: all dangerous and experimental features are enabled.
2019-02-27 11:52:02.550 7f2c6ed7a700 -1 WARNING: all dangerous and experimental features are enabled.
dumped pgs
PG_STAT OBJECTS MISSING_ON_PRIMARY DEGRADED MISPLACED UNFOUND BYTES OMAP_BYTES* OMAP_KEYS* LOG DISK_LOG STATE        STATE_STAMP                VERSION REPORTED UP      UP_PRIMARY ACTING  ACTING_PRIMARY LAST_SCRUB SCRUB_STAMP                LAST_DEEP_SCRUB DEEP_SCRUB_STAMP           SNAPTRIMQ_LEN 
1.9          16                  0        0         0       0     0   201326592         48  16       16 active+clean 2019-02-27 07:08:43.559261   13'16    16:40 [1,2,0]          1 [1,2,0]              1      13'16 2019-02-26 14:22:42.286676           13'16 2019-02-26 14:22:42.286676             0 
1.8          21                  0        0         0       0     0   264241152         63  21       21 active+clean 2019-02-27 07:08:43.554860   13'21    16:45 [1,2,0]          1 [1,2,0]              1      13'21 2019-02-26 14:22:41.355824           13'21 2019-02-26 14:22:41.355824             0 
1.7          39                  0        0         0       0     0   490733568        117  39       39 active+clean 2019-02-27 07:08:43.545804   13'39    16:63 [1,2,0]          1 [1,2,0]              1      13'39 2019-02-26 14:22:40.473369           13'39 2019-02-26 14:22:40.473369             0 
1.6          35                  0        0         0       0     0   440401920        105  35       35 active+clean 2019-02-27 07:08:43.556879   13'35    16:59 [1,0,2]          1 [1,0,2]              1      13'35 2019-02-26 14:22:36.503362           13'35 2019-02-26 14:22:36.503362             0 
1.1          28                  0        0         0       0     0   352321536         84  28       28 active+clean 2019-02-27 07:08:43.535766   13'28    16:52 [2,0,1]          2 [2,0,1]              2      13'28 2019-02-26 14:22:23.320667           13'28 2019-02-26 14:22:23.320667             0 
1.0          23                  0        0         0       0     0   289406976         69  23       23 active+clean 2019-02-27 07:08:43.559662   13'23    16:47 [1,0,2]          1 [1,0,2]              1      13'23 2019-02-26 14:22:30.478020           13'23 2019-02-26 14:22:30.478020             0 
1.2          38                  0        0         0       0     0   478150656        114  38       38 active+clean 2019-02-27 07:08:43.539575   13'38    16:62 [0,1,2]          0 [0,1,2]              0      13'38 2019-02-26 14:22:24.268516           13'38 2019-02-26 14:22:24.268516             0 
1.3          27                  0        0         0       0     0   339738624         81  27       27 active+clean 2019-02-27 07:08:43.552369   13'27    16:51 [1,2,0]          1 [1,2,0]              1      13'27 2019-02-26 14:22:31.512259           13'27 2019-02-26 14:22:31.512259             0 
1.4          32                  0        0         0       0     0   402653184         96  32       32 active+clean 2019-02-27 07:08:43.554496   13'32    16:56 [1,0,2]          1 [1,0,2]              1      13'32 2019-02-26 14:22:35.549114           13'32 2019-02-26 14:22:35.549114             0 
1.5          37                  0        0         0       0    24   452984832        108  37       37 active+clean 2019-02-27 07:08:43.536722   13'37    16:61 [2,0,1]          2 [2,0,1]              2      13'37 2019-02-26 14:22:26.340739           13'37 2019-02-26 14:22:26.340739             0 

* NOTE: Omap statistics are gathered during deep scrub and may be inaccurate soon afterwards depending on utilisation. See http://docs.ceph.com/docs/master/dev/placement-group/#omap-statistics for further details.
```

**Comment by badone** [comment]
```
$ ceph pg ls
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-02-27 11:52:30.822 7f67d10a7700 -1 WARNING: all dangerous and experimental features are enabled.
2019-02-27 11:52:30.980 7f67d10a7700 -1 WARNING: all dangerous and experimental features are enabled.
PG  OBJECTS DEGRADED MISPLACED UNFOUND BYTES OMAP_BYTES* OMAP_KEYS* LOG STATE        SINCE VERSION REPORTED UP        ACTING    SCRUB_STAMP                DEEP_SCRUB_STAMP           
1.0      23        0         0       0     0   289406976         69  23 active+clean    4h   13'23    16:47 [1,0,2]p1 [1,0,2]p1 2019-02-26 14:22:30.478020 2019-02-26 14:22:30.478020 
1.1      28        0         0       0     0   352321536         84  28 active+clean    4h   13'28    16:52 [2,0,1]p2 [2,0,1]p2 2019-02-26 14:22:23.320667 2019-02-26 14:22:23.320667 
1.2      38        0         0       0     0   478150656        114  38 active+clean    4h   13'38    16:62 [0,1,2]p0 [0,1,2]p0 2019-02-26 14:22:24.268516 2019-02-26 14:22:24.268516 
1.3      27        0         0       0     0   339738624         81  27 active+clean    4h   13'27    16:51 [1,2,0]p1 [1,2,0]p1 2019-02-26 14:22:31.512259 2019-02-26 14:22:31.512259 
1.4      32        0         0       0     0   402653184         96  32 active+clean    4h   13'32    16:56 [1,0,2]p1 [1,0,2]p1 2019-02-26 14:22:35.549114 2019-02-26 14:22:35.549114 
1.5      37        0         0       0    24   452984832        108  37 active+clean    4h   13'37    16:61 [2,0,1]p2 [2,0,1]p2 2019-02-26 14:22:26.340739 2019-02-26 14:22:26.340739 
1.6      35        0         0       0     0   440401920        105  35 active+clean    4h   13'35    16:59 [1,0,2]p1 [1,0,2]p1 2019-02-26 14:22:36.503362 2019-02-26 14:22:36.503362 
1.7      39        0         0       0     0   490733568        117  39 active+clean    4h   13'39    16:63 [1,2,0]p1 [1,2,0]p1 2019-02-26 14:22:40.473369 2019-02-26 14:22:40.473369 
1.8      21        0         0       0     0   264241152         63  21 active+clean    4h   13'21    16:45 [1,2,0]p1 [1,2,0]p1 2019-02-26 14:22:41.355824 2019-02-26 14:22:41.355824 
1.9      16        0         0       0     0   201326592         48  16 active+clean    4h   13'16    16:40 [1,2,0]p1 [1,2,0]p1 2019-02-26 14:22:42.286676 2019-02-26 14:22:42.286676 

* NOTE: Omap statistics are gathered during deep scrub and may be inaccurate soon afterwards depending on utilisation. See http://docs.ceph.com/docs/master/dev/placement-group/#omap-statistics for further details.
```

File: src/mon/PGMap.cc (hunk 1 above)
Comment (badone) [reply src/mon/PGMap.cc]: Will do.

Comment (badone) [reply src/mon/PGMap.cc]: Actually, doing it this way just eats my new line between the output and my note so I might stick with the original @gregsfortytwo

**Comment by neha-ojha** [comment]
jenkins render docs

**Comment by badone** [comment]
@gregsfortytwo will provide a follow-up commit with a test and per osd output but didn't want to risk this not making Nautilus.

**Pull request merged**